src/AggregateFunctions/AggregateFunctionSparkbar.h

```diff
@@ -0,0 +1,310 @@
+     *  For example, there are actually 11 buckets, specify 10 buckets, and divide the 11 buckets as follows (11/10):
+     *  0.0-1.1, 1.1-2.2, 2.2-3.3, 3.3-4.4, 4.4-5.5, 5.5-6.6, 6.6-7.7, 7.7-8.8, 8.8-9.9, 9.9-11.
+     *  The y value of the first bucket will be calculated as follows:
+     *  the actual y value of the first position + the actual second position y*0.1, and the remaining y*0.9 is reserved for the next bucket.
+     *  The next bucket will use the last y*0.9 + the actual third position y*0.2, and the remaining y*0.8 will be reserved for the next bucket. And so on.
+     */
+    String render(const AggregateFunctionSparkbarData<X, Y> & data) const
+    {
+        String value;
+        if (data.points.empty())
+            return value;
+        X local_min_x = data.points.begin()->first;
+        X local_max_x = data.points.rbegin()->first;
+        size_t diff_x = local_max_x - local_min_x;
+        if ((diff_x + 1) <= width)
+        {
+            Y min_y = data.min_y;
+            Y max_y = data.max_y;
+            Float64 diff_y = max_y - min_y;
+
+            if (diff_y)
+            {
+                for (size_t i = 0; i <= diff_x; ++i)
+                {
+                    auto it = data.points.find(local_min_x + i);
+                    bool found = it != data.points.end();
+                    value += getBar(found ? std::round(((it->second - min_y) / diff_y) * 7) + 1 : 0);
+                }
+            }
+            else
+            {
+                for (size_t i = 0; i <= diff_x; ++i)
+                    value += getBar(data.points.count(local_min_x + i) ? 1 : 0);
+            }
+        }
+        else
+        {
+            // begin reshapes to width buckets
+            Float64 multiple_d = (diff_x + 1) / static_cast<Float64>(width);
+
+            // init bounds
+            std::map<size_t, Float64> bounds;
+            for (size_t i = 1; i <= width; ++i)
+            {
+                Float64 bound = i * multiple_d;
+                bounds.template emplace(std::floor(bound), bound);
+            }
+
+            std::optional<Float64> min_y;
+            std::optional<Float64> max_y;
+
+            std::optional<Float64> new_y;
+            std::map<size_t, std::optional<Float64>> newPoints;
+            for (size_t i = 0; i <= (diff_x + 1); ++i)
+            {
+                auto bound = bounds.find(i);
+                if (bound != bounds.end()) // is bound
+                {
+                    Float64 proportion = bound->second - bound->first;
+                    auto it = data.points.find(local_min_x + i);
+                    bool found = (it != data.points.end());
+                    if (found)
+                        new_y ? new_y = new_y.value() + it->second * proportion : new_y = it->second * proportion;
+
+                    // find bucket_num
+                    size_t bucket_num = 0;
+                    for (auto & b : bounds)
```

Comment:
Isn't `bucket_num = std::floor(i / multiple_d)`? And if it's true, do we need `bounds` at all?

Reply:
Thanks.